import { container } from "tsyringe";
import { ILogger } from "@coreShared/logs/ILogger";
import { LoggerContext } from "@coreShared/logs/LoggerContext";
import { ResultType } from "@coreShared/types/result.type";
import util from "util";

function safe(data: any) {
    try {
        JSON.stringify(data);
        return data;
    } catch {
        return util.inspect(data, { depth: 2 });
    }
}

export function LogExecution(): MethodDecorator {
    return function (
        target: any,
        propertyKey: string | symbol,
        descriptor: PropertyDescriptor
    ) {
        const originalMethod = descriptor.value;

        descriptor.value = async function (...args: any[]) {
            const logger = container.resolve<ILogger>("ILogger");
            const context = new LoggerContext(target.constructor.name, propertyKey.toString());
            const start = Date.now();

            try {
                await logger.logInfo(context, "üöÄ Iniciando execu√ß√£o");

                const result = await originalMethod.apply(this, args);

                const duration = Date.now() - start;

                if (result instanceof ResultType) {
                    if (result.isSuccess()) {
                        await logger.logInfo(
                            context,
                            `‚úîÔ∏è Execu√ß√£o conclu√≠da em ${duration}ms`,
                            undefined,
                            // { result: safe(result) }
                        );
                    } else if( result.isNone()) {
                        await logger.logInfo(
                            context,
                            `‚úîÔ∏è Execu√ß√£o conclu√≠da (nenhum resultado) em ${duration}ms`,
                            undefined,
                            // { result: safe(result) }
                        );
                    }
                    else {
                        const error = result.getError();
                        await logger.logWarn(
                            context,
                            `‚ö†Ô∏è Erro de neg√≥cio: ${error.message ?? error}`,
                            undefined,
                            // { error: safe(error), duration }
                        );
                    }
                } else {
                    await logger.logInfo(
                        context,
                        `‚úîÔ∏è Execu√ß√£o conclu√≠da (sem ResultType) em ${duration}ms`,
                        undefined,
                        // { result: safe(result) }
                    );
                }

                return result;
            } catch (error: any) {
                const duration = Date.now() - start;
                await logger.logError(
                    context,
                    `‚ùå Erro inesperado: ${error.message ?? error}`,
                    error.stack,
                    { duration }
                );
                throw error;
            }
        };
    };
}
